[clang] Catch missing format attributes

budimirarandjelovichtec opened the pull request:
Enable flag -Wmissing-format-attribute to catch missing attributes.

Fixes https://github.com/llvm/llvm-project/issues/60718

budimirarandjelovichtec committed:
[clang] Catch missing format attributes

budimirarandjelovichtec commented:
ping @AaronBallman @aaronpuchert @vitalybuka @Endilll

AaronBallman commented:
Can you explain what changes were made to address the various issues which caused the earlier revert? I've tried diffing the PRs to see what the changes are, but my git-fu is insufficient to the task. :-D

vitalybuka commented:
> Can you explain what changes were made to address the various issues which caused the earlier revert? I've tried diffing the PRs to see what the changes are, but my git-fu is insufficient to the task. :-D

Just for convenience I extracted update diff into https://github.com/llvm/llvm-project/pull/106649

budimirarandjelovichtec commented:
> Can you explain what changes were made to address the various issues which caused the earlier revert? I've tried diffing the PRs to see what the changes are, but my git-fu is insufficient to the task. :-D

I made changes in several places. Here are major changes:
1) determining first argument index. Here I removed dynamic cast to DeclRefAttr of last function argument. In C++ mode casting to DeclRefAttr resulted in unrecognizing argument type as va_list for several architectures and OS. So, some architectures gave 0 as output and others gave non-zero value. Unrecognizing argument type as va_list was a result of defining canonical va_list type differently depending on architecture.
2) added check if calling function has format attribute before iterating through its format attributes. This check was added because one pointer (AttrType) was initialized in iteration through format attributes and later used to get name if missing attribute was caught. In previous merge this was resulting in throwing error that there were use of uninitialized variable for some architectures.
3) edited some parts of C code where passed arguments were not valid. Examples are passing &args[0] or char * as va_list argument (functions f8() and f41()).

Relevant parts of code are reviewed in [#106649 ](https://github.com/llvm/llvm-project/pull/106649/).

AaronBallman commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -5335,6 +5335,221 @@ static void handlePreferredTypeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+    unsigned int ChildFunctionFormatArgumentIndexOffset =
+        checkIfMethodHasImplicitObjectParameter(ChildFunction) ? 2 : 1;
+
+    if (!ChildFunction->hasAttr<FormatAttr>())
+      continue;
```
```suggestion
```

AaronBallman commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -5335,6 +5335,221 @@ static void handlePreferredTypeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+      continue;
+    const FunctionDecl *ChildFunction =
+        dyn_cast_or_null<FunctionDecl>(TheCall->getCalleeDecl());
+    if (!ChildFunction)
```
```suggestion
    if (!ChildFunction || !ChildFunction->hasAttr<FormatAttr>())
```

AaronBallman commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -5335,6 +5335,221 @@ static void handlePreferredTypeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+    int FirstToCheck = [&]() -> unsigned {
+      if (!FDecl->isVariadic())
+        return 0;
+      const auto *FirstToCheckArg = Args[NumArgs - 1]->IgnoreParenCasts();
+      if (!FirstToCheckArg ||
+          FirstToCheckArg->getType().getCanonicalType() !=
+              Context.getBuiltinVaListType().getCanonicalType())
+        return 0;
```
This change doesn't make sense to me -- you explained in [this comment](https://github.com/llvm/llvm-project/pull/106649/files#r1738137575) (thank you for the comments on that diff, btw!) that the cast to `DeclRefExpr` caused problems, but this code is equivalent (`DeclRefExpr` has no `getType()` method, it inherits it from `Expr`). I don't think the original cast was necessary, but I don't know that this will have fixed the issue.

If this code is in fact correct, there's no need to check `!FirstToCheckArg` because there's no way for that to be a null pointer now.

budimirarandjelovichtec replied:
Result of Args[NumArgs - 1]->IgnoreParenCasts() points to a memory location in case of all architectures that I tested. However, casting to DeclRefExpr could produce nullptr depending on architecture. So, for some architectures DeclRefExpr * type pointed to a memory address and for some not. Then I decided to remove casting.

AaronBallman replied:
Oh! I see now, okay, that makes more sense, thank you! Then I think you can remove the `!FirstToCheckArg`.

budimirarandjelovichtec replied:
Removed ```!FirstToCheckArg```

AaronBallman commented on clang/test/Sema/attr-format-missing.c:
```diff
@@ -0,0 +1,393 @@
+void f10(const wchar_t *out, ... /* args */) // #f10
+{
+    va_list args;
+    vscanf(out, args);
```
This looks wrong -- `vscanf` takes a `const char *` as the first argument, I think you want `vwscanf` instead, right?

AaronBallman commented on clang/test/Sema/attr-format-missing.c:
```diff
@@ -0,0 +1,393 @@
+void f11(const wchar_t *out, ... /* args */) // #f11
+{
+    va_list args;
+    vscanf((const char *) out, args); // expected-warning@#f11 {{diagnostic behavior may be improved by adding the 'scanf' format attribute to the declaration of 'f11'}}
```
Same here as above?

AaronBallman replied:
This code makes no sense; that conversion is still UB.

AaronBallman commented on clang/test/Sema/attr-format-missing.c:
```diff
@@ -0,0 +1,393 @@
+void f12(const wchar_t *out, ... /* args */) // #f12
+{
+    va_list args;
+    vscanf((char *) out, args); // expected-warning@#f12 {{diagnostic behavior may be improved by adding the 'scanf' format attribute to the declaration of 'f12'}}
```
Here too.

budimirarandjelovichtec replied:
Here is also a conversion.

AaronBallman commented on clang/test/Sema/attr-format-missing.c:
```diff
@@ -0,0 +1,393 @@
+void f13(const wchar_t *out, ... /* args */) // #f13
+{
+    va_list args;
+    vscanf(out, args);
```
and here

AaronBallman commented on clang/test/Sema/attr-format-missing.c:
```diff
@@ -0,0 +1,393 @@
+                       // CHECK: fix-it:"{{.*}}":{[[@LINE-4]]:6-[[@LINE-4]]:6}:"__attribute__((format(scanf, 1, 0)))"
+}
+
+void f15(const char16_t *out, ... /* args */) // #f15
```
I don't think you need to test f15 through f25, they're all type mismatches with the first argument to `vscanf` and there are no functions with the appropriate signature for the types to replace them with (unlike with `wchar_t` where there is a library API we can use).

budimirarandjelovichtec replied:
I will remove functions that passes non-char type as first argument. Is it worth to keep functions where are conversions of first argument to char * (eg. f18, f19, f20)?

AaronBallman replied:
Nope, I think those should go as well.

AaronBallman commented on clang/test/Sema/attr-format-missing.cpp:
```diff
@@ -0,0 +1,174 @@
+void f5(const std::wstring &str, ... /* args */) // #f5
+{
+    va_list args;
+    vprintf((const char *)str.c_str(), args); // expected-no-warning@#f5
```
Use `vwprintf` instead of casting.

AaronBallman commented on clang/test/Sema/attr-format-missing.cpp:
```diff
@@ -0,0 +1,174 @@
+void f7(std::wstring_view str, ... /* args */) // #f7
+{
+    va_list args;
+    vprintf((const char *) std::wstring(str).c_str(), args); // expected-no-warning@#f7
```
Same here.

AaronBallman reviewed (approved):
LGTM!

budimirarandjelovichtec commented:
Let's wait few days if there are any opinions. Then, someone from my company will merge it.

aaronpuchert commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -5335,6 +5335,217 @@ static void handlePreferredTypeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+    int FirstToCheck = [&]() -> unsigned {
+      if (!FDecl->isVariadic())
+        return 0;
+      const auto *FirstToCheckArg = Args[NumArgs - 1]->IgnoreParenCasts();
+      if (FirstToCheckArg->getType().getCanonicalType() !=
```
Why do we check only the last argument against `va_list`? Is there a rule that it cannot come in another position?

budimirarandjelovichtec replied:
I overlooked that va_list can be also in another position. Originally I added this check to determine first-to-check index depending if last argument is va_list. However, now it seems to me that first-to-check index can be determined by only checking if caller function is variadic.
As passing va_list to variadic function results in UB, I added a check at the begin of iteration. If callee function is variadic, iteration exits.
Also, in the same part of code I added check if va_list is passed to callee function.

aaronpuchert commented on clang/test/Sema/attr-format-missing.cpp:
```diff
@@ -0,0 +1,176 @@
+{
+    va_list args;
+    fn2(out, args); // expected-warning@#S1_fn3 {{diagnostic behavior may be improved by adding the 'scanf' format attribute to the declaration of 'fn3'}}
+                    // CHECK: fix-it:"{{.*}}":{[[@LINE-4]]:10-[[@LINE-4]]:10}:"__attribute__((format(scanf, 2, 3)))"
```
Shouldn't we suggest adding the attribute to the first (in-class) declaration of the function instead of the definition here?

budimirarandjelovichtec replied:
Fixed.

budimirarandjelovichtec commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -5335,6 +5335,217 @@ static void handlePreferredTypeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+    // If it is, check if parent function misses format attributes.
+
+    unsigned int ChildFunctionFormatArgumentIndexOffset =
+        checkIfMethodHasImplicitObjectParameter(ChildFunction) ? 2 : 1;
```
Changed names to caller and callee.

budimirarandjelovichtec commented on clang/test/Sema/attr-format-missing.c:
```diff
@@ -0,0 +1,251 @@
+// RUN: %clang_cc1 -fsyntax-only -Wmissing-format-attribute -fdiagnostics-parseable-fixits %s 2>&1 | FileCheck %s
+// RUN: %clang_cc1 -fsyntax-only -x c++ -verify=expected,cpp_diagnostics -Wmissing-format-attribute %s
+// RUN: %clang_cc1 -fsyntax-only -x c++ -verify=expected,cpp_diagnostics -std=c++2b -Wmissing-format-attribute %s
+// RUN: %clang_cc1 -fsyntax-only -x c++ -verify=expected,cpp_diagnostics -std=c++23 -Wmissing-format-attribute %s
```
I googled regarding this. Didn't find that c++2b and c++23 differ. So, I removed c++2b.

Endilll replied:
Yes, `c++2b` and `c++23` refer to the same language mode. I think when `c++23` was added, all uses of `c++2b` in the tests were replaced. We should be consistent with that.

budimirarandjelovichtec commented on clang/test/Sema/attr-format-missing.cpp:
```diff
@@ -0,0 +1,176 @@
+
+    void fn4(this S1& expliciteThis, const char *out, va_list args) // #S1_fn4
+    {
+        expliciteThis.fn2(out, args); // beforeCxx2b-error@#S1_fn4 {{explicit object parameters are incompatible with C++ standards before C++2b}}
```
Guarded with ```cxx_explicit_this_parameter```.

budimirarandjelovichtec commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -5335,6 +5335,217 @@ static void handlePreferredTypeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+        return 0;
+      const auto *FirstToCheckArg = Args[NumArgs - 1]->IgnoreParenCasts();
+      if (FirstToCheckArg->getType().getCanonicalType() !=
+          Context.getBuiltinVaListType().getCanonicalType())
```
BuiltinVaListType isn't canonical. Type of getBuiltinVaListType() is '__builtin_va_list'. Type of getBuiltinVaListType().getCanonicalType() is 'struct __va_list_tag[1]'.
Btw, I modified and moved this check to the begin of iteration.

vitalybuka reviewed (changes requested):
Before looking into details, may I ask you to:
1. Re-base
2. address or reply to existing comments
3. click re-request review.

Thank you!

budimirarandjelovichtec commented:
@vitalybuka
Yes, I will address existing comments.
I was on PTO previous week and now I'm continuing to work on this PR.

budimirarandjelovichtec commented on clang/test/Sema/attr-format-missing.c:
```diff
@@ -0,0 +1,251 @@
+{
+    va_list args;
+
+    vscanf(out, args); // expected-no-warning@#f8
+    vprintf(out, args); // expected-no-warning@#f8
```
@AaronBallman WDYT?

budimirarandjelovichtec replied:
@AaronBallman Do you agree with @aaronpuchert or stick for not emitting diagnostics?

budimirarandjelovichtec commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -5335,6 +5335,230 @@ static void handlePreferredTypeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+                                 .getBuiltinVaListType()
+                                 .getCanonicalType()
+                                 .getBaseTypeIdentifier()
+                                 ->getName()) {
```
I found that on some architectures builtin_va_list canonical type is char *. In order make program capable to distinguish real builtin_va_list and other char *, I added this comparison. As this does not give appropriate results, I returned previous comparison.

budimirarandjelovichtec commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -5335,6 +5335,217 @@ static void handlePreferredTypeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+  for (auto *Child : Body->children()) {
+    // If child statement is compound statement, recursively get missing
+    // attributes.
+    if (dyn_cast_or_null<CompoundStmt>(Child)) {
```
The idea of this part of code is to enable diagnosing missing format attributes if there are branches around call, ie. call is not direct child of caller function. This part of code is aimed to handle examples like this:

```
void f(char *out, va_list args)
{
    {
        vprintf(out, args);
    }
}
```

Without this check, missing format attribute in example above would not be diagnosed. This part of code enables recursively checking children statements.

If statement is not compound, next check is if it is a function call.

aaronpuchert replied:
When you write about "branches around call", do you mean braces? However, I meant branches:
```c++
void f(char *out, va_list args)
{
    if (tracing_enabled) {
        vprintf(out, args);
    }
}
```
This is not a compound statement, so you'd not be looking into that.

My proposal would be to check on encountering an annotated `CallExpr` instead of traversing the function body. This is also better for performance.

budimirarandjelovichtec replied:
Yes, I meant braces instead of branches. This snippet does not produce warning, so I'm working to fix it.

budimirarandjelovichtec replied:
Changed logic to detect missing format attributes on an annotated function calls. Here are differences: [link](https://github.com/llvm/llvm-project/compare/97b70712c658bae20657dd78253d8de10319bf2d..68fb91662185dc164574f0058e660310c05e08a4).

budimirarandjelovichtec commented on clang/test/Sema/attr-format-missing.c:
```diff
@@ -0,0 +1,251 @@
+void f1(char *out, const size_t len, const char *format, ... /* args */) // #f1
+{
+    va_list args;
+    vsnprintf(out, len, format, args); // expected-no-warning@#f1
```
It seems that ```-verify``` does not handle 'expected-no-warning' comment, so I disabled this comment.

Endilll replied:
> It seems that `-verify` does not handle 'expected-no-warning' comment, so I disabled this comment.

It doesn't need to. If there's no expected directive on the line, it is interpreted as if you're putting a hypothetical `expected-no-warning` there.
If you wanted to make it clear for human readers, you can leave something along the lines of `// OK` on that line.

budimirarandjelovichtec replied:
Thank you for explanation!

budimirarandjelovichtec commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -5335,6 +5335,217 @@ static void handlePreferredTypeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+    }
+
+    // Do not add in a vector format attributes whose type is different than
+    // parent function attribute type.
```
Snippet moved to ```DiagnoseMissingFormatAttributes```.

budimirarandjelovichtec commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -5335,6 +5335,217 @@ static void handlePreferredTypeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+      return NumOfParentFunctionParams + FunctionFormatArgumentIndexOffset;
+    }();
+
+    // If there are already attributes which arguments matches arguments
+    // detected in this iteration, do not add new attribute as it would be
+    // duplicate.
+    if (!llvm::any_of(MissingAttributes, [&](const FormatAttr *Attr) {
+          return Attr->getType() == AttrType &&
+                 Attr->getFormatIdx() == StringIndex &&
+                 Attr->getFirstArg() == FirstToCheck;
+        }))
```
It is true that duplicates are also eliminated earlier, but keep in mind that this iteration forks and can go in one of three opportunities:
1) If child statement is compound statement, format attributes are got recursively for child statement and it this branch is one elimination of duplicates. Then, loop continues to next iteration.
2) If child statement is function call, there is check if format attribute is missing. If it is case, before adding to the vector, there is check for duplicates. After all, loop continues to next iteration.
3) If none of previous conditions are satisfied, loop continues to next iteration.

So, parts of code that eliminate duplicates are in different branches and both snippets cannot be executed in one iteration.

aaronpuchert commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -5335,6 +5335,217 @@ static void handlePreferredTypeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+                        return true;
+                      })) {
+      MissingAttributes.push_back(
+          FormatAttr::CreateImplicit(getASTContext(), AttrType, -1, -1));
```
But one of those calls might not be forwarding the parameter, and thus be irrelevant:
```c++
void f(char *in, va_list args)
{
    printf("Accepting input: ");
    vscanf(out, args);
}
```
Didn't try it, but it looks like this might not emit the warning because of `printf`. And forwarding to both `printf` and `scanf` doesn't really make sense to me, or at least it's probably rare enough that we don't need to worry about it.

budimirarandjelovichtec replied:
I tested this function and it emits diagnostic due to va_list checking (snippet below). As va_list is not passed to printf, format type info is not saved, so diagnostic is emitted. If this snippet was removed, there would be no warning.

```
    // Check if va_list is passed to callee function.
    // If va_list is not passed, continue to the next iteration.
    bool hasVaList = false;
    for (const auto *Param : CalleeFunction->parameters()) {
      if (Param->getOriginalType().getCanonicalType() ==
          S.getASTContext().getBuiltinVaListType().getCanonicalType()) {
        hasVaList = true;
        break;
      }
    }
    if (!hasVaList)
      continue;
```

aaronpuchert replied:
Fair enough, but I think it's too complicated. Passing the `va_list` to multiple functions with different attributes should be extremely rare, and I don't see how it can make sense. So warning about it is probably fine.

budimirarandjelovichtec commented on clang/test/Sema/attr-format-missing.c:
```diff
@@ -0,0 +1,251 @@
+void f2(char *out, const size_t len, const char *format, ... /* args */) // #f2
+{
+    va_list args;
+    vsnprintf(out, len, format, args); // expected-warning@#f2 {{diagnostic behavior may be improved by adding the 'printf' format attribute to the declaration of 'f2'}}
```
Yes, warning is emitted on the declaration. I've just added a note to the call that trigerred it.

budimirarandjelovichtec commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -5335,6 +5335,217 @@ static void handlePreferredTypeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+    // Check if format string argument is parent function parameter.
+    int StringIndex = 0;
+    if (!llvm::any_of(FDecl->parameters(), [&](const ParmVarDecl *Param) {
+          if (Param != FormatArg)
```
As format string argument is determined in previous ```any_of``` and it seems that this snippet is redundant, I removed it. Also, I changed names of ```FormatArg``` and ```StringIndex``` to ```FormatString(Arg|Index)```.

budimirarandjelovichtec commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -5335,6 +5335,217 @@ static void handlePreferredTypeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+                return false;
+
+              if (FDecl->parameters()[OffsetFormatIndex] != FormatArg)
+                return false;
```
In previous, ```any_of``` it was found ```callee``` format argument. In this ```any_of```, ```caller``` format argument is checked with ```callee``` one. If they mismatch, it means that ```caller``` misses format attribute.

budimirarandjelovichtec commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -5335,6 +5335,230 @@ static void handlePreferredTypeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+
+    // va_list is not intended to be passed to variadic function.
+    if (CalleeFunction->isVariadic())
+      continue;
```
Also, can you check previous comments where my response is last?

budimirarandjelovichtec replied:
On the other hand, this example does not emit diagnostics.
```
__attribute__((format(printf, 1, 0)))
void f1X(const char *out, int a, double b)
{
    printf(out, a, b);
}
```

budimirarandjelovichtec replied:
ping @aaronpuchert

budimirarandjelovichtec commented:
> Before looking into details, may I ask you to:
> 
>     1. Re-base
> 
>     2. address or reply to existing comments
> 
>     3. click re-request review.
> 
> 
> Thank you!

1) Rebased to the last commit as of February 14, 2025.

2) Replied on existing comments.

3) Re-requested review for all reviewers.

budimirarandjelovichtec commented:
Ping @vitalybuka @aaronpuchert

apple-fcloutier commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -5918,6 +5918,181 @@ static void handlePreferredTypeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+// vector. Vector could contains attributes that only store information about
+// format type (format string and first to check argument are set to -1).
+namespace {
+std::vector<FormatAttr *> MissingAttributes;
```
Clang does not use global mutable state. You need to find some other way to keep that around.

apple-fcloutier reviewed (changes requested):
As I understand your change, it can only trip when -Wformat-nonliteral also trips. I think that you should implement this as a note/fixit on -Wformat-nonliteral. You should be able to leverage `checkFormatString`'s value traversal to have the same precision as -Wformat checking. I don't think that your change correctly identifies this case, for instance:

```c
void foo(const char *a, ...) {
	va_list ap;
	const char *const b = a;
	va_start(ap, a);
	vprintf(b, ap);
	va_end(ap);
}
```

whereas -Wformat will dig that deep.

aaronpuchert reviewed (commented):
> As I understand your change, it can only trip when `-Wformat-nonliteral` also trips. I think that you should implement this as a note/fixit on `-Wformat-nonliteral`.

Good observation, and I agree. It's even documented that this is one way to fix the warning:

> Clang implements this mostly the same way as GCC, but there is a difference for functions that accept a `va_list` argument (for example, `vprintf`). GCC does not emit `-Wformat-nonliteral` warning for calls to such functions. Clang does not warn if the format string comes from a function parameter, where the function is annotated with a compatible attribute, otherwise it warns.

An interesting question is whether we should have a separate warning about forwarding that doesn't warn about other nonliterals. But I don't see why one would want that. After all the idea of these warnings is to follow format strings all the way up the stack until we reach a constant string that we can check.

This could also simplify the change a lot.
